Serialize trash correctly

johnswanson opened the pull request:
Two major changes here.

First, don't set the trash ID to a constant 13371339. Instead, just add it like a normal collection (in a migration), with an appropriate `type`, and cache that so we're not constantly fetching the ID when needed.

Second, serialize the Trash correctly. If we're serializing a set of collections, serialize any items that were _trashed from_ one of those collections, as well as items that are still in the collection. However, hardcode the `descendants` of the Trash collection itself to the empty set. This way, even though we're serializing the Trash, we won't cascade down and serialize all of its descendants - just those that are being otherwise included in the export.

Fixes #42301

johnswanson committed:
Remove `ensure-trash-collection-created!`

This is no longer necessary, as we're creating the trash collection in a
migration rather than on startup.

johnswanson committed:
Fix audit tests

johnswanson committed:
Don't copy parent permissions for trashed colls

johnswanson committed:
Don't set a specific trash collection.id

johnswanson committed:
Serialize the Trash correctly

johnswanson committed:
fix downgrade migration

johnswanson committed:
Fix deserialization of trashed collections

A `nil` `trashed_from_parent_id` can indicate either that the collection
was trashed from the root collection, or that the collection isn't
archived at all. When setting `trashed_from_location`, use the
`archived` value to distinguish between the two possibilites.

johnswanson committed:
Fix a few tests

These were failing due to the existence of the Trash collection
post-migration.

noahmoss commented on enterprise/backend/src/metabase_enterprise/serialization/cmd.clj:
```diff
@@ -236,12 +237,16 @@
+        ;; we _ALWAYS_ export the Trash. Its descendants are empty, so we won't export anything extra as a result, but
+        ;; we will export items that are currently in the Trash, assuming they were trashed *from* a place we're
+        ;; exporting.
+        collection-ids+trash (set/union collection-ids #{(collection/trash-collection-id)})
```
nit - inconsistent alignment

noahmoss commented on enterprise/backend/test/metabase_enterprise/serialization/v2/e2e_test.clj:
```diff
@@ -478,15 +480,17 @@
 
-                  (is (= [{:id                   "abc",
+                  ;; card1s has no parameters, card2s does.
+                  (is (= #{[]
+                           [{:id                   "abc",
```
nit - inconsistent alignment

noahmoss commented on enterprise/backend/test/metabase_enterprise/serialization/v2/extract_test.clj:
```diff
@@ -44,6 +45,8 @@
        (map :id)
        set))
 
+(def trash-eid (delay (:entity_id (collection/trash-collection))))
```
can this be private?

noahmoss commented on enterprise/backend/test/metabase_enterprise/serialization/v2/storage_test.clj:
```diff
@@ -70,18 +76,19 @@
-                  p-dir  (str (:entity_id parent)      "_parent_collection")
-                  c-dir  (str (:entity_id child)       "_child_collection")]
-              (is (= #{[gp-dir (str gp-dir ".yaml")]                                          ; Grandparent collection
+            (let [gp-dir    (str (:entity_id grandparent) "_grandparent_collection")
+                  p-dir     (str (:entity_id parent)      "_parent_collection")
+                  c-dir     (str (:entity_id child)       "_child_collection")]
```
nit - extra whitespace. same below.

noahmoss commented on resources/migrations/001_update_migrations.yaml:
```diff
@@ -6929,6 +6929,30 @@ databaseChangeLog:
+        - sql:
+            sql: >-
+              INSERT INTO collection (name, slug, entity_id, type) VALUES ('Trash', 'trash', 'trashtrashtrashtrasht', 'trash');
+              INSERT INTO permissions (object, group_id)
```
This made me wonder — do we ensure that the trash collection's permissions can't be modified via the API?

johnswanson replied:
Yep, we do: https://github.com/metabase/metabase/blob/c6da3f9a5cc4c471511167b2520edeeaa92b0ac0/src/metabase/models/permissions.clj#L523-L524

noahmoss commented on src/metabase/models/collection.clj:
```diff
@@ -46,40 +45,37 @@
-  (memoize trash-collection*))
-
-(def trash-path
+(defn- trash-collection* []
+  (t2/select-one :model/Collection :type "trash")
+  )
```
nit
```suggestion
(defn- trash-collection* []
  (t2/select-one :model/Collection :type "trash"))
```

noahmoss commented on src/metabase/models/collection.clj:
```diff
@@ -1050,57 +1064,74 @@
-                           (or (:entity_id parent) (serdes/identity-hash parent)))
-        owner-email      (when (:personal_owner_id coll)
-                           (t2/select-one-fn :email 'User :id (:personal_owner_id coll)))]
+  (let [fetch-collection       (fn [id]
+                              (t2/select-one Collection :id id))
```
nit - indentation

noahmoss commented on src/metabase/models/collection.clj:
```diff
@@ -1050,57 +1064,74 @@
+                parent_id]}    (some-> coll :id fetch-collection (t2/hydrate :parent_id :trashed_from_parent_id))
+        parent                 (some-> parent_id fetch-collection)
+        trashed-from-parent    (some-> trashed_from_parent_id fetch-collection)
+        parent-id              (when parent
+                              (or (:entity_id parent) (serdes/identity-hash parent)))
```
nit - indentation

noahmoss reviewed (approved):
just nits - sorry for harping on the whitespace inconsistencies, I'm just sensitive to them since my editor will fix (some of) them automatically when I open the file, leading to unexpected unstaged changes 😆

johnswanson committed:
Fix CR nits

johnswanson committed:
Don't set Trash permissions for admins

We don't have any other permissions set in the DB for administrators -
they can do anything they want and don't need to have specific
permissions rows.

johnswanson commented:
Thanks @noahmoss for noticing all those nits! 😅
